src/text/text_buffer.rs

```diff
@@ -122,6 +122,14 @@ impl<R: BufRead> TextBuffer<R> {
             // The invariants enforced by the `consume` method guarantee that this will behave safely.
```

Comment:
🗺️ 

> // The `copy_within()` method below is unsafe.
> // ....
> // The invariants enforced by the `consume` method guarantee that this will behave safely.

Famous last words. 🤦

Once in a blue moon, this method would lead to a panic. Here's what happened.

`TextBuffer` has a `String` that acts as its reusable buffer. Any time we need more data from STDIN, it calls `read_line()` some number of times and appends the new text to the end of the string. When the reader finds an incomplete value at the end of the buffer, it shifts that remaining text to the beginning of the buffer and then reads more in. This works fine.

What I hadn't anticipated was the way that `String::truncate` works. When you truncate a `String` to a new length (`new_len`), [it checks to make sure that the byte at `new_len` is a character boundary](https://doc.rust-lang.org/src/alloc/string.rs.html#1202). What's surprising about this (at least to me) is that `new_len` is _beyond_ the end of the buffer after truncation. `String` is effectively looking at here-be-dragons data that used to have meaning. 🐉 

I was doing the `is_char_boundary` check _before_ shifting the remaining text backwards in the buffer and also each time that the reader called `consume` to flag bytes as having been read. This was functionally the same thing that `truncate` was doing. However, I never imagined that it would be checking the value of a byte that was not in the `String`'s readable range.

On rare occasions (like when reading in data that used surrogate pairs), the buffer would be filled with multi-byte code points. My `is_char_boundary` checks worked great, but after shifting the remaining text backwards in the buffer, the byte at `new_len` would be in the middle of what used to be a multi-byte code point. Kaboom.

This fixes the issue by setting the byte at `new_len` to a zero.

Anyway, that's what I get for using an `unsafe` method. 😞